mixer/adapter/circonus/circonus.go

```diff
@@ -0,0 +1,171 @@
+		cm          cgm.CirconusMetrics
+		metricTypes map[string]*metric.Type
+		env         adapter.Env
+		metrics     map[string]config.Params_MetricInfo_Type
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &handler{}
+
+///////////////// Configuration-time Methods ///////////////
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+
+	cmc := &cgm.Config{}
+	cmc.CheckManager.Check.SubmissionURL = b.adpCfg.SubmissionUrl
+	cmc.Debug = true
+	cmc.Interval = "0"
+	cm, err := cgm.NewCirconusMetrics(cmc)
+	if err != nil {
+		err = env.Logger().Errorf("Could not create NewCirconusMetrics: %v", err)
+		return nil, err
+	}
+
+	metrics := make(map[string]config.Params_MetricInfo_Type)
+	ac := b.adpCfg
+	for _, metric := range ac.Metrics {
+
+		metrics[metric.Name] = metric.Type
+	}
+	return &handler{cm: *cm, metricTypes: b.metricTypes, env: env, metrics: metrics}, nil
+}
+
+// adapter.HandlerBuilder#SetAdapterConfig
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adpCfg = cfg.(*config.Params)
+}
+
+// adapter.HandlerBuilder#Validate
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	return nil
+}
+
+// metric.HandlerBuilder#SetMetricTypes
+func (b *builder) SetMetricTypes(types map[string]*metric.Type) {
+	b.metricTypes = types
+}
+
+////////////////// Request-time Methods //////////////////////////
+// metric.Handler#HandleMetric
+func (h *handler) HandleMetric(ctx context.Context, insts []*metric.Instance) error {
+	var result *multierror.Error
+
+	for _, inst := range insts {
+
+		metricName := inst.Name
+		if _, ok := h.metrics[metricName]; !ok {
+			result = multierror.Append(result, fmt.Errorf("Cannot find Type for instance %s", metricName))
+			continue
+		}
+
+		metricType, found := h.metrics[metricName]
+		if !found {
+			result = multierror.Append(result, fmt.Errorf("no type for metric named %s", metricName))
+			continue
+		}
+
+		switch metricType {
+
+		case config.GAUGE:
+
+			value, ok := inst.Value.(int64)
+
+			if !ok {
+				result = multierror.Append(result, fmt.Errorf("could not record gauge '%v': %v, %v", metricName, inst, value))
+				continue
+			}
+
+			h.cm.Gauge(metricName, value)
+
+		case config.COUNTER:
+
+			value, ok := inst.Value.(int64)
+
+			if !ok {
+				result = multierror.Append(result, fmt.Errorf("could not record counter '%s': %v, %v", metricName, inst.Value, value))
+				continue
+			}
+
+			h.cm.Increment(metricName)
+
+		case config.DISTRIBUTION:
+
+			v, ok := inst.Value.(time.Duration)
+
+			if ok {
+				h.cm.Timing(metricName, float64(v))
+				continue
+			}
+
+			vint, ok := inst.Value.(float64)
+			if ok {
+				h.cm.Timing(metricName, float64(vint))
+				continue
+			}
+			result = multierror.Append(result, fmt.Errorf("could not record distribution '%s': %v", metricName, inst.Value))
+
+		}
+
+	}
+	h.env.ScheduleWork(h.cm.Flush)
```

Comment:
It would be better if the interaction with the backend were handled synchronously so that errors are reported up. As it is, errors in this flush code will be invisible to the rest of the Istio pipeline.

Reply:
I was able to tie the circonus-gometrics logger which uses the standard log package into the `env.Logger()` interface. Log entries are being reported with the async approach for flush events. Let me know if this handles the spirit of what you were after here.